From e163ff16c34e66e7c5b3434a1fe8e7074c65fc2b Mon Sep 17 00:00:00 2001
From: Wang ShaoBo <bobo.shaobowang@huawei.com>
Date: Mon, 29 Nov 2021 16:28:37 +0800
Subject: [PATCH] arm64/mpam: Fix use-after-free in mkdir_resctrl_prepare()
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: e163ff16c34e66e7c5b3434a1fe8e7074c65fc2b


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I4K272
CVE: NA

-------------------------------------------------

We add label out_free_rdtgrp for handling error branch when it happened
before rmid and closid allocation, in case of reusing rdtgrp after freeing.

Fixes: 0b16164dc7a9 ("arm64/mpam: Remap reqpartid,pmg to rmid and intpartid to closid")
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/resctrlfs.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/fs/resctrlfs.c b/fs/resctrlfs.c
index e1c996ff4e79..7ca9fe3ee4a4 100644
--- a/fs/resctrlfs.c
+++ b/fs/resctrlfs.c
@@ -749,7 +749,7 @@ static int mkdir_resctrl_prepare(struct kernfs_node *parent_kn,
 		ret = closid_alloc();
 		if (ret < 0) {
 			rdt_last_cmd_puts("out of CLOSIDs\n");
-			goto out_unlock;
+			goto out_free_rdtgrp;
 		}
 		rdtgrp->closid.intpartid = ret;
 	}
@@ -819,10 +819,11 @@ static int mkdir_resctrl_prepare(struct kernfs_node *parent_kn,
 	kernfs_remove(rdtgrp->kn);
 out_free_rmid:
 	rmid_free(rdtgrp->mon.rmid);
-	kfree(rdtgrp);
 out_free_closid:
 	if (rdtgrp->type == RDTCTRL_GROUP)
 		closid_free(rdtgrp->closid.intpartid);
+out_free_rdtgrp:
+	kfree(rdtgrp);
 out_unlock:
 	resctrl_group_kn_unlock(prgrp_kn);
 	return ret;
-- 
2.26.2

